README.md
# ldap_nis

#### Table of Contents

1. [Description](#description)
1. [Setup - The basics of getting started with ldap_nis](#setup)
    * [Installing the gem](#Installing the gems)
    * [Providing LDAP credentials](#Providing LDAP credentials)
    * [Beginning with ldap_nis](#beginning-with-ldap_nis)
1. [Usage - Configuration options and additional functionality](#usage)
1. [References](#reference)
  * [Types](#Types)
    * [ldap_entity](#ldap_entity) - a native type for creating arbitrary ldap entries
    * [ldap_nis::domain](#ldap_nis::domain) - domain object
    * [ldap_nis::host](#ldap_nis::host) - host entry
    * [ldap_nis::user](#ldap_nis::user) - user entry
    * [ldap_nis::group](#ldap_nis::group) - group entry
    * [ldap_nis::netgroup](#ldap_nis::netgroup) - netgroup entry
    * [ldap_nis::automountmap](#ldap_nis::automountmap) - automount map (auto_master, auto_direct, etc.)
    * [ldap_nis::automountentry](#ldap_nis::automountentry) - automount map mountpoint entry
    * [ldap_nis::mailgroup](#ldap_nis::mailgroup) - mail group aliases
    * [ldap_nis::network](#ldap_nis::network) - network entry
    * [ldap_nis::service](#ldap_nis::service) - ip services
  * [Functions](#Functions)
    * [domain2dn](#domain2dn)
    * [sha1digest](#sha1digest)
1. [Limitations](#limitations)
1. [Developmente](#development)

## Description

This module centers around an LDAP type/provider that allows creation of arbitrary ldap entries using the ruby net-ldap gem.  Layered on top of this is a series of defined types that support creation of NIS style entries for user, group, hosts, and more.

The use of the net-ldap gem means that any host which realizes LDAP entries must have that gem installed.  The provider detects the presence of this gem and will fail if undetected.  The gem is ldap implementation neutral and should work against any ldap compliant instance.

## Setup

### Installing the gem

Make sure the hosts creating ldap entries have access to the net-ldap gem.  This may be accomplished using the package gem provider.

    package { 'net-ldap':
      ensure   => present,
      provider => puppet_gem,
    }

### Providing LDAP credentials
While you may provide credentials for each resource definition it is easier to specify this information globally using hiera.

    ldap_nis::server::base:     'dc=yourdomain,dc=net'
    ldap_nis::server::host:     'localhost'
    ldap_nis::server::username: 'cn=admin,dc=yourdomain,dc=net'
    ldap_nis::server::password: 'plaintext'
    ldap_nis::server::port:     636
    ldap_nis::server::ssl:      true
    ldap_nis::server::ssl_cacert: '/etc/ssl/certs/myca-bundle.crt'
    ldap_nis::server::verify:   true


* `base` - root of the NIS object which holds the containers for the various NIS entries (ou=people, ou=hosts, etc.)
* `host` - hostname or ip of ldap server
* `username` - account with write privileges for the `base` structure
* `password` - associated password
* `port` - either the unencrypted port (389) or the ldaps:/// encrypted port (636) when using SSL
* `ssl` - whether or not to use ldaps:/// SSL encryption
* `ssl_cacert` - Path to a PEM file with your trusted CAs.
* `verify` - whether or not to verify the ssl certificate and/or server DNS entry

The `ssl` option only supports the ldaps:// interface on an ecrypted port (usually 636).  When using SSL you may choose to disable certificate validation by setting `verify` to false.  This disables all verification and opens up the possibility for man-in-the-middle attacks.  By default, the puppet net-ldap gem uses a private CA bundle to authenticate certificates.  This makes it impossible to simply add your certificate to the OS trust store.  The `ssl_cacert` allows you to specify your own list CA trusted certificates.

### Beginning with ldap_nis

The first object to create is the NIS domain object.  The following example is for a generic domain used my "yourdomian.net" and assumes the credentials listed above.

    $dns_name = 'yourdomain.net'
    $base     = domain2dn($dns_name)
    $domain   = 'yourdomain'

    ldap_nis::domain { $dns_name:
      ensure => present,
      dc     => $domain,
    }

This will create the NIS domain object and by default will create supporting containers (ou=people, ou=group, etc.).  You may now create additional entries for users, groups, netgroups, etc.

## Usage
### Attribute Mutability

It may be desirable to allow certain attributes to change after creation.  Attributes like passwords may be changed by the user or as a result of hashing during creation.  Marking an attribute mutable by including it in the `mutable` array causes the attribute to accept its specified value at creation but to otherwise be ignored.

## Reference

### Types
* [ldap_entity](#ldap_entity) - a native type for creating arbitrary ldap entries
* [ldap_nis::domain](#ldap_nis-domain) - domain object
* [ldap_nis::host](#ldap_nis::host) - host entry
* [ldap_nis::user](#ldap_nis::user) - user entry
* [ldap_nis::group](#ldap_nis::group) - group entry
* [ldap_nis::netgroup](#ldap_nis::netgroup) - netgroup entry
* [ldap_nis::automountmap](#ldap_nis::automountmap) - automount map (auto_master, auto_direct, etc.)
* [ldap_nis::automountentry](#ldap_nis::automountentry) - automount map mountpoint entry
* [ldap_nis::mailgroup](#ldap_nis::mailgroup) - mail group aliases
* [ldap_nis::network](#ldap_nis::network) - network entry
* [ldap_nis::service](#ldap_nis::service) - ip services

#### ldap_entity
Allows creation of raw ldap entries.  Here is an example of manually creating a domain entry for yourdomain.net:

    ldap_entity { 'dc=yourdomain,dc=net':
      ensure     => present,
      base       => 'dc=yourdomain,dc=net',
      host       => 'localhost',
      port       => '636'
      ssl        => true,
      verify     => true,
      username   => 'cn=admin,dc=yourdomain,dc=net',
      password   => 'plaintext',
      attributes => {
        objectclass  => [ 'top', 'domain', 'nisdomainobject' ],
        dc           => 'yourdomain',
        nisdomain    => 'yourdomain.net'
      }
    }

Then you would need to create each of the containers within the domain with entries like:

    ldap_entity { 'ou=people,dc=yourdomain,dc=net':
      ensure     => present,
      base       => 'dc=yourdomain,dc=net',
      host       => 'localhost',
      port       => '636'
      ssl        => true,
      verify     => true,
      username   => 'cn=admin,dc=yourdomain,dc=net',
      password   => 'plaintext',
      attributes => {
        objectclass  => [ 'top', 'organizationalunit' ],
        ou           => 'people',
      }
    }

#### ldap_nis::domain
DN format: `"${base}"`

Here is the entry to create the domain using the provided defined type:

    ldap_nis::domain { 'yourdomain.net':
      ensure => present,
      dc     => 'yourdomain',
    }

The provider creates the containers by default unless you specify:

  create_containers => false,

#### ldap_nis::host
DN format: `"cn=${name}+ipHostNumber=${iphostnumber},ou=hosts,${base}"`

    ldap_nis::host { 'myhost.yourdomain.net'
      ensure       => present,
      iphostnumber => '1.1.1.1',
    }

| Parameter       | Default | Description                  |
| :-------------- | :-----: | :--------------------------- |
| `name`          | `title` | The name should be the fully qualified domain name.  This is the primary name for the host and is what is returned for reverse ip lookups. Defaults to the titlebar value if not specified   |
|  `iphostnumber` |         | IP address of host            |
|  aliases        | []      | An array of hostname aliases |

#### ldap_nis::user
DN format: `"uid=${name},ou=${container},${base}"`

    ldap_nis::user { 'myuser':
      ensure => present,
      uidnumber     => '1000',
      gidnumber     => '1000',
      sn            => 'myuser',
      homedirectory => '/home/myuser',
    }

| Parameter       | Default | Description                  |
| :-------------- | :-----: | :--------------------------- |
| `name`          | `title` | Username for entry           |
| `uidnumber`     |         | UserID number                |
| `gidnumber`     |         | GroupID number               |
| `sn`            |         | Surname of the user          |
| `homedirectory` |         | Users home directory         |
| loginshell      | undef   | Path to the loginshell       |
| container       | people  | OU container for account.  Allows placing users into different containers.  By default **people** and **application** containers are provided. |
| employeenumber  | undef    | Employee number of the user |
| userpassword    | undef    | Password for the user       |
| gecos           | undef    | GECOS information for the user |

#### ldap_nis::group
DN format: `"cn=${name},ou=group,${base}"`

    ldap_nis::group { 'mygroup':
      gidnumber => 1000,
    }

| Parameter       | Default | Description                  |
| :-------------- | :-----: | :--------------------------- |
| `name`          | `title` | Group Name                   |
| `gidnumber`     |         | GroupID number               |
| gr_password     | undef   | Group Password               |
| members         | undef   | Array of member UIDs         |
| description     | undef   | Description of group         |

#### ldap_nis::netgroup
DN format: `"cn=${name},ou=netgroup,${base}`

    ldap_nis::netgroup { 'mynetgroup':
      ensure => present,
    }

| Parameter       | Default | Description                  |
| :-------------- | :-----: | :--------------------------- |
| `name`          | `title` | Name of netgroup             |
| members         | []      | Array of netgroup members which may include individual netgroup triples or the names of other netgroups |

#### ldap_nis::automountmap

DN format: `"automountmapname=${name},${base}"`

    ldap_nis::automountmap { 'my_map':
      ensure => present,
    }

| Parameter       | Default | Description                  |
| :-------------- | :-----: | :--------------------------- |
| `name`          | `title` | Name of Automount Map        |
| description     | undef   | Description for Automount Map |

#### ldap_nis::automountentry
DN format: `"automountkey=${name},automountmapname=${mapname},${base}"`

    ldap_nis::automountentry { '/my/applications':
      mapname => 'my_map',
      info    => '-rw,bg,retry=2,soft,vers=3 myappserver:/myapps',
    }

| Parameter       | Default | Description                  |
| :-------------- | :-----: | :--------------------------- |
| `name`          | `title` | Automount entry mountpoint directory|
| `mapname`       |         | Name of map for this entry   |
| `info`          |         | Mount information for entry  |
| description     | undef   | Description for this entry   |

#### ldap_nis::mailgroup
DN format: `"mail=${name},ou=aliases,${base}""`

    ldap_nis::mailgroup { 'mymailgroup':
      ensure => present,
      members => [ 'user1', 'user2' ],
    }

| Parameter       | Default | Description                  |
| :-------------- | :-----: | :--------------------------- |
| `name`          |         | Mail group name              |
| `members`       |         | Array of mailgroup member addresses |

#### ldap_nis::network
DN format: `"ipnetworknumber=${name},ou=networks,${base}"`

    ldap_nis::networks { '1.0.0.0':
      label       => 'RED1',
      description => 'The Big Red 1',
    }

| Parameter       | Default | Description                  |
| :-------------- | :-----: | :--------------------------- |
| `name`          | `title` | Network portion of address space |
| `label`         |         | Short name for network       |
| ipnetmask       | undef   | Netmask for network (CIDR Addressing) |
| description     | undef   | Description of network       |

#### ldap_nis::service
DN format: `"cn=${name}+ipserviceprotocol=${ipserviceprotocol},ou=services,${base}`

    ldap_nis { 'mytcpport':
      ipservicerport    => '1000',
      ipserviceprotocol => 'tcp',
    }

| Parameter       | Default | Description                  |
| :-------------- | :-----: | :--------------------------- |
| `name`          | `title` | IP Port Name                 |
| `ipserviceport` |         | IP Port Number               |
| `ipserviceprotocol` |     | IP Protocol                  |
| description     | undef   | Description of IP Port       |

## Functions

#### Hash a password with SHA-1 Digest

```ruby
sha1digest("secret") # => "{SHA}5en6G6MezRroT3XKqkdPOmY/BfQ="
```

#### Convert a dotted domain to a DN format suitable for LDAP

```ruby
domain2dn("yourdomain.net") # => "dc=yourdomain,dc=net"
```


## Limitations

The first iteration of this module assumes the use of an ldap server with solaris schema extensions installed.  These extensions for OpenLDAP are available in the files directory.  It is my intention to support other configurations in the future and time and energy permits.

## Development

Since your module is awesome, other users will want to play with it. Let them
know what the ground rules for contributing are.

## Release Notes/Contributors/Etc.

This module began life as an extension to the datacentred/ldap module and morphed into a general purpose ldap entry management tool.  The ldap_entity type/provider is a lightly overhauled version of the ldap_entry type found in that module.

### Running tests

<!--- TODO - need to work on tests --->
This project contains tests uses [rspec-puppet](http://rspec-puppet.com/) to verify functionality. Test coverage is minimal at best at this point, but should improve with time.

Quickstart:

    gem install bundler
    bundle install
    bundle exec rake spec


CHANGELOG.md
##  Release 0.1.0 - 2017-04-15
#### Summary
Initial release with support for Solaris NIS Schema

Features
- Native Type/Provider ldap_entity supports arbitrary ldap entries
- Defined Types for Core NIS entries
  - ldap_nis::domain
  - ldap_nis::host
  - ldap_nis::user
  - ldap_nis::group
  - ldap_nis::netgroup
  - ldap_nis::automountmap
  - ldap_nis::automountentry
  - ldap_nis::mailgroup
  - ldap_nis::network
- Parser Functions
  - domain2dn
  - sha1digest


spec/classes/init_spec.rb
require 'spec_helper'
describe 'ldap_nis' do
  context 'with default values for all parameters' do
    it { should contain_class('ldap_nis') }
  end
end


lib/puppet/provider/ldap_entity/net_ldap.rb
require 'rubygems' if RUBY_VERSION < '1.9.0' && Puppet.version < '3'
require 'net/ldap' if Puppet.features.net_ldap?
require 'net/ldap/dn' if Puppet.features.net_ldap?

Puppet::Type.type(:ldap_entity).provide(:ldap) do
  confine :feature => :net_ldap

  public

  def attributes=(value)
    value.each do |k, v|
      if @property_hash[:attributes].has_key?(k)
        # skip if current value matches new
        next if [@property_hash[:attributes][k]].flatten.sort == [v].flatten.sort

        # skip if attribute is marked as mutable
        next if [resource[:mutable]].flatten.include? k
        ldap_replace_attribute([resource[:host], resource[:port], resource[:username], resource[:password],
                    [resource[:name], k, v]])
      else
        ldap_add_attribute([resource[:host], resource[:port], resource[:username], resource[:password],
                      [resource[:name], k, v]])
      end
    end
    @property_hash[:attributes]=value
  end
  #
  def attributes
    @property_hash[:attributes]
  end

  def exists?
    status, result = ldap_search([resource[:host], resource[:port],
      resource[:username], resource[:password], {
        :base => resource[:name], :scope => Net::LDAP::SearchScope_BaseObject,
        :attributes => attributes_keys(resource[:attributes])
      }])
    if status == LDAP_NIS::NoSuchObject
      Puppet.debug("LDAP exists? expect #{resource[:ensure]==:absent ? false :
            true}; #{resource[:name]} is false")
      return false
    elsif status == LDAP_NIS::Success
      if Net::LDAP::DN.new(result.dn).to_a.map(&:downcase) ==
                Net::LDAP::DN.new(resource[:name]).to_a.map(&:downcase)

        # We have a match, store attributes
        entry = Hash.new
        result.each_attribute do |attribute, value|
          next if attribute == :dn

          # Force ojectclass values to lowercase
          value.map!(&:downcase) if attribute == :objectclass

          # Store values
          entry[attribute.to_s.downcase]=value
        end
        @property_hash[:attributes] = entry
        Puppet.debug("LDAP exists? expect #{resource[:ensure]==:absent ? false : true} #{resource[:name]} is true")
        return true
      else
        Puppet.debug("LDAP exists? expect #{resource[:ensure]==:absent ? false : true} #{resource[:name]} is false")
        return false
      end
    else
      raise "LDAP Error #{status}: #{result}. Check server log for more info."
    end
  end

  def destroy
    Puppet.info("LDAP destroy #{resource[:name]}")
    resource[:ensure] = :absent
    status, message = ldap_remove([resource[:host], resource[:port], resource[:username], resource[:password],
          {:dn => resource[:name]}])
    raise "LDAP Error #{status}: #{message}. Check server log for more info." unless status == LDAP_NIS::Success
  end

  def create
    Puppet.info("LDAP create #{resource[:name]}")
    status, message = ldap_add([resource[:host], resource[:port], resource[:username], resource[:password],
                        {:dn => resource[:name], :attributes => resource[:attributes]}])
    raise "LDAP Error #{status}: #{message}. Check server log for more info." unless status == LDAP_NIS::Success
  end

  private

  def ldap(args)
    host, port, admin_user, admin_password, _ = args

    tls_options={}
    tls_options[:ca_file] = resource[:ssl_cacert] unless resource[:ssl_cacert].empty?
    tls_options[:verify_mode] = OpenSSL::SSL::VERIFY_NONE unless resource[:verify]

    ldap = Net::LDAP.new(
      { :host => host,
        :port => port,
        :auth => {
          :method => :simple,
          :username => admin_user,
          :password => admin_password}}.
          merge(resource[:ssl] ? {:encryption => {
            :method => :simple_tls,
            :tls_options => tls_options}} : {}))
    Puppet.debug("Connecting to LDAP server ldaps://#{host}:#{port}")
    ldap.bind
    ldap
  end

  def attributes_keys(attrs)
    return [] unless resource[:attributes]
    attrs.keys.map(&:to_s)
  end

  def ldap_search(args)
    ldap = ldap(args)
    Puppet.debug("LDAP Search: #{params(args).inspect}")
    args = params(args)
    results = ldap.search(args)
    code, message = return_code_and_message(ldap)

    if(code == LDAP_NIS::Success)
      [code, results.first]
    else
      [code, message]
    end
  end

  def ldap_add(args)
    ldap = ldap(args)
    Puppet.info("LDAP Add: #{params(args).inspect}")
    ldap.add(params(args))
    return_code_and_message(ldap)
  end

  def ldap_remove(args)
    ldap = ldap(args)
    Puppet.info("LDAP Remove: #{params(args).inspect}")
    ldap.delete(params(args))
    return_code_and_message(ldap)
  end

  def ldap_add_attribute(args)
    ldap = ldap(args)
    Puppet.info("LDAP Add Attribute: #{params(args).inspect}")
    ldap.add_attribute(*params(args))
    return_code_and_message(ldap)
  end

  def ldap_replace_attribute(args)
    ldap = ldap(args)
    Puppet.info("LDAP Replace Attribute: #{params(args).inspect}")
    ldap.replace_attribute(*params(args))
    return_code_and_message(ldap)
  end

  def params(args)
    args[4]
  end

  def return_code_and_message(ldap)
    result = ldap.get_operation_result
    return result.code, LDAP_NIS.lookup_error(result.code)
  end

  # For more, see http://www.zytrax.com/books/ldap/ch12/
  module LDAP_NIS
    Success                      = 0
    OperationsError              = 1
    ProtocolError                = 2
    TimeLimitExceeded            = 3
    SizeLimitExceeded            = 4
    CompareFalse                 = 5
    CompareTrue                  = 6
    StrongAuthNotSupported       = 7
    StrongAuthRequired           = 8
    OnlyPartialResultsReturned   = 9
    LdapReferral                 = 10
    AdminLimitExceeded           = 11
    UnavailableCriticalExtension = 12
    ConfidentialityRequired      = 13
    SaslBindInProgress           = 14
    NoSuchAttribute              = 16
    UndefinedAttributeType       = 17
    InappropriateMatching        = 18
    ConstraintViolation          = 19
    AttributeOrValueExists       = 20
    InvalidSyntax                = 21
    NoSuchObject                 = 32
    AliasProblem                 = 33
    InvalidDNSyntax              = 34
    ObjectIsLeaf                 = 35
    AliasDereferenceProblem      = 36
    InappropriateAuthentication  = 48
    InvalidCredentials           = 49
    InsufficientAccessRights     = 50
    Busy                         = 51
    Unavailable                  = 52
    UnwillingToPerform           = 53
    LoopDetected                 = 54
    NamingViolation              = 64
    ObjectClassViolation         = 65
    NotAllowedOnNonLeaf          = 66
    NotAllowedOnRDN              = 67
    EntryAlreadyExists           = 68
    NoObjectClassModifications   = 69
    ResultsTooLarge              = 70
    AffectsMultipleDSAs          = 71
    UnknownError                 = 80

    def self.lookup_error(code)
      Hash[constants.collect{|c| [
        LDAP_NIS.const_get(c), c.to_s.gsub(/([A-Z])/, ' \1').strip
      ]}][code]
    end
  end
end


lib/puppet/type/ldap_entity.rb
require 'rubygems' if RUBY_VERSION < '1.9.0' && Puppet.version < '3'
require 'net/ldap' if Puppet.features.net_ldap?

Puppet::Type.newtype(:ldap_entity) do
  @doc = 'Type to manage LDAP entities'

  ensurable

  newparam(:name) do
    desc 'Name of LDAP entity i.e. cn=foo,ou=bar,dc=sampel,dc=com'
    isnamevar
  end

  newparam(:host) do
    desc 'Host Address (FQDN or IP) of the LDAP server'
  end

  newparam(:base) do
    desc 'LDAP tree base i.e. dc=foo,dc=co,dc=uk'
    validate do |value|
      unless value.is_a? String
        raise ArgumentError, 'ldap_entity::base is not a string'
      end
    end
  end

  newparam(:port) do
    desc  'Port of the LDAP server (default 389)'
    defaultto 636
    validate do |value|
      unless (1..65535).include?(value.to_i)
        raise ArgumentError, 'ldap_entity::port is not a whole number in the range 1-65535'
      end
    end
  end

  newparam(:username) do
    desc 'Username of admin account on LDAP server'
    defaultto 'admin'
    validate do |value|
      unless value.is_a? String
        raise ArgumentError, 'ldap_entity::username is not a string'
      end
    end
  end

  newparam(:password) do
    desc 'Password of admin account on LDAP server'
    validate do |value|
      unless value.is_a? String
        raise ArgumentError, 'ldap_entity::password is not a string'
      end
    end
  end

  newproperty(:attributes) do
    desc 'LDAP entry attributes as a hash i.e. { :givenName => "Foo",
                                                 :objectClass =>
                                                   ["top", "person", "inetorgPerson"]}'
    validate do |value|
      unless value.is_a? Hash
        raise ArgumentError, 'ldap_entity::attributes is not a hash'
      end

      value.each_pair do |k,v|
        if k[/^[[:upper:]]*$/]
          raise ArgumentError, 'ldap_entity::attributes keys not all lowercase'
        end
        if k == "objectclass"
          raise ArgumentError, 'ldap_entity::attributes - objectclass must be an array' unless v.kind_of?(Array)
          v.each do |entry|
            raise ArgumentError, 'ldap_entity::attributes - objectclass values must be lowercase' if entry[/^[[:upper:]]*$/]
          end
        end
      end
    end

    def insync?(is)
      should.each do |k,v|
        return false unless [is[k]].flatten.sort == [should[k]].flatten.sort
      end
      return true
    end
  end

  newparam(:mutable, :array_matching => :all) do
    desc 'LDAP entry attribute(s) as an array which may be externally modified and therefore unamaged beyond the initial value'
    defaultto []

    validate do |value|
      unless value.is_a? Array
        raise ArgumentError, 'ldap_entity::mutable is not an array'
      end
    end
  end

  newparam(:verify) do
    desc 'Should the LDAP server certificate and host be verified'
    defaultto true
  end

  newparam(:ssl) do
    desc 'Whether the LDAP server uses SSL'
    defaultto true
  end

  newparam(:ssl_cacert) do
    desc 'Path to a CA file used to verify SSL certificates'
    defaultto String.new('')
  end

  # Add autorequire
  autorequire(:ldap_entry) do
    # Strip off the first dn to autorequire the parent
    parent = self[:name].split(",").drop(1).join(",")
    parent
  end

end
